HW2/P2/parallel_vector.pyx

```diff
@@ -107,9 +118,22 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+           for idx in prange(src.shape[0], num_threads=4, schedule=dynamic):
+
+                if counts[src[idx]] > 0:
+                   if src[idx]/N == dest[idx]/N:
+                       acquire(&locks[dest[idx]/N])
+                       counts[dest[idx]] += 1
+                       counts[src[idx]] -= 1
+                       release(&locks[dest[idx]/N])
+
+                   else:
+                       acquire(&locks[dest[idx]/N])
+                       counts[dest[idx]] += 1
+                       release(&locks[dest[idx]/N])
+
+                       acquire(&locks[src[idx]/N])
+                       counts[src[idx]] -= 1
+                       release(&locks[src[idx]/N])
 
```

Comment:
Hm nice clean loops, I had many ifs/elseifs in this part that were kind of annoying. I did that because I think you have to impose some sort of order for this problem, seems like you might've got around that?